File: validator-core/src/main/resources/validator.js
function SimpleDateFormat(pattern) {
	var object = new Object();
	object.pattern = pattern;
	object.regex = new RegExp("^" + pattern.replace("yyyy", "\\d{4}").replace("MM", "(0\\d|1[12])").replace("dd", "([0-2]\\d|3[0-1])")
		.replace("HH", "([0-1]\\d|2[0-3])").replace("hh", "(0\\d|1[0-2])").replace("mm", "[0-5]\\d").replace("ss", "[0-5]\\d") + "$");
	object.position = {
		year: pattern.indexOf("yyyy"), month: pattern.indexOf("MM"), day: pattern.indexOf("dd"),
		hour: pattern.toLowerCase().indexOf("hh"), minute: pattern.indexOf("mm"), second: pattern.indexOf("ss")
	};
	object.parse = function(source) {
		if(!this.regex.test(source))
			throw new Error("Unparseable date: \"" + source + "\"");
		var time = {
			year: source.substr(this.position.year, 4), month: source.substr(this.position.month, 2), day: source.substr(this.position.day, 2)
		};
		if (this.position.hour != -1)
			time.hour = source.substr(this.position.hour, 2);
		if (this.position.minute != -1)
			time.minute = source.substr(this.position.minute, 2);
		if (this.position.second != -1)
			time.second = source.substr(this.position.second, 2);
		var day31 = "01,03,05,07,08,10,12";
		if(time.day == 31 && day31.indexOf(time.month) == -1)
			throw new Error("Unparseable date: \"" + source + "\"");
		if(time.month == 2 && time.day == 29 && !(time.year % 4 == 0 && time.year % 100 != 0)
			&& !(time.year % 100 == 0 && time.year % 400 == 0)) {
			throw new Error("Unparseable date: \"" + source + "\"");
		}
		var date = new Date();
		date.setFullYear(time.year, time.month - 1, time.day);
		if (time.hour != undefined) date.setHours(time.hour);
		if (time.minute != undefined) date.setMinutes(time.minute);
		if (time.second != undefined) date.setSeconds(time.second);
		return date;
	};
	object.format = function(date) {
		function fmt(v,n){for(var i=n-(v+"").length;i>0;i--){v="0"+v;}return v;}
		var h24 = date.getHours();
		return this.pattern.replace("yyyy", fmt(date.getFullYear(), 4)).replace("MM", fmt(date.getMonth() + 1, 2))
			.replace("dd", fmt(date.getDate(), 2)).replace("HH", fmt(h24, 2)).replace("hh", fmt((h24-1)%12+1, 2))
			.replace("mm", fmt(date.getMinutes(), 2)).replace("ss", fmt(date.getSeconds(), 2));
	};
	return object;
}
function Validator(selector, handlerError) {
	var valid = new Object();
	valid.form = $(selector);
	valid.invalid = false;
	valid.shortCircuit = false;
	valid.datePattern = null;
	valid.error = {};
	valid.checkedField = null;
	valid.handlerError = handlerError;
	var DEFAULT_DATE_PATTERN = "yyyy-MM-dd";
	var emailAddressPattern = "\\b(^['_A-Za-z0-9-]+(\\.['_A-Za-z0-9-]+)*@([A-Za-z0-9-])+(\\.[A-Za-z0-9-]+)*((\\.[A-Za-z0-9]{2,})|(\\.[A-Za-z0-9]{2,}\\.[A-Za-z0-9]{2,}))$)\\b";
	var mobileNumberPattern = "^(0|86|17951)?(13[0-9]|15[0-9]|17[678]|18[0-9]|14[57])[0-9]{8}$";
	var emailAddressManyPattern = "^((['_A-Za-z0-9-]+(\\.['_A-Za-z0-9-]+)*@([A-Za-z0-9-])+(\\.[A-Za-z0-9-]+)*((\\.[A-Za-z0-9]{2,})|(\\.[A-Za-z0-9]{2,}\\.[A-Za-z0-9]{2,}));)+)$";
	var urlAddressPattern = "";
	
	valid.setShortCircuit = function(shortCircuit) {
		valid.shortCircuit = shortCircuit;
	};
	valid.setDatePattern = function(datePattern) {
		valid.datePattern = datePattern;
	};
	valid.getDatePattern = function() {
		return (valid.datePattern != null ? valid.datePattern : DEFAULT_DATE_PATTERN);
	};
	valid.validate = function () {};
	valid.addError = function(errorKey, errorMessage) {
		valid.invalid = true;
		if(valid.error[errorKey] == null)
			valid.error[errorKey] = errorMessage;
		if (valid.shortCircuit) {
			throw new Error();
		}
	};
	valid.getParameter = function(name) {
		return valid.form.find("[name="+name+"]").val();
	};
	valid.setCheckedField = function(field) {
		valid.checkedField = field;
	};
	valid.isCheckedField = function(field) {
		return valid.checkedField == null || field == valid.checkedField;
	};
	valid.validateRequired = function(field, errorKey, errorMessage) {
		if (!this.isCheckedField(field)) return;
		delete valid.error[errorKey];
		var value = this.getParameter(field);
		if (value == null || "" == value)
			this.addError(errorKey, errorMessage);
	};
	valid.validateRequiredString = function(field, errorKey, errorMessage) {
		if (!this.isCheckedField(field)) return;
		delete valid.error[errorKey];
		var value = this.getParameter(field);
		if (value == null || "" == value.replace(/\s/ig, ""))
			this.addError(errorKey, errorMessage);
	};
	valid.validateInteger = function(field, min, max, errorKey, errorMessage) {
		if (!this.isCheckedField(field)) return;
		delete valid.error[errorKey];
		var value = this.getParameter(field);
		if (value == null || "" == value.replace(/\s/ig, "")) {
			this.addError(errorKey, errorMessage);
			return ;
		}
		var temp = parseInt(value.replace(/\s/ig, ""));
		if (isNaN(temp) || temp < min || temp > max)
			this.addError(errorKey, errorMessage);
	};
	valid.validateInteger = function(field, errorKey, errorMessage) {
		if (!this.isCheckedField(field)) return;
		delete valid.error[errorKey];
		var value = this.getParameter(field);
		if (value == null || "" == value.replace(/\s/ig, "")) {
			this.addError(errorKey, errorMessage);
			return ;
		}
		if(isNaN(parseInt(value.replace(/\s/ig, "")))) {
			this.addError(errorKey, errorMessage);
			return ;
		}
	};
	valid.validateLong = function(field, min, max, errorKey, errorMessage) {
		if (!this.isCheckedField(field)) return;
		delete valid.error[errorKey];
		this.validateInteger(field, min, max, errorKey, errorMessage);
	};
	valid.validateLong = function(field, errorKey, errorMessage) {
		if (!this.isCheckedField(field)) return;
		delete valid.error[errorKey];
		this.validateInteger(field, errorKey, errorMessage);
	};
	valid.validateDouble = function(field, min, max, errorKey, errorMessage) {
		if (!this.isCheckedField(field)) return;
		delete valid.error[errorKey];
		var value = this.getParameter(field);
		if (value == null || "" == value.replace(/\s/ig, "")) {
			this.addError(errorKey, errorMessage);
			return ;
		}
		var temp = parseFloat(value.replace(/\s/ig, ""));
		if (isNaN(temp) || temp < min || temp > max)
			this.addError(errorKey, errorMessage);
	};
	valid.validateDouble = function(field, errorKey, errorMessage) {
		if (!this.isCheckedField(field)) return;
		delete valid.error[errorKey];
		var value = this.getParameter(field);
		if (value == null || "" == value.replace(/\s/ig, "")) {
			this.addError(errorKey, errorMessage);
			return ;
		}
		if (isNaN(parseFloat(value.replace(/\s/ig, ""))))
			this.addError(errorKey, errorMessage);
	};
	valid.validateDate = function(field, errorKey, errorMessage) {
		if (!this.isCheckedField(field)) return;
		delete valid.error[errorKey];
		var value = this.getParameter(field);
		if (value == null || "" == value.replace(/\s/ig, "")) {
			this.addError(errorKey, errorMessage);
			return ;
		}
		try {
			new SimpleDateFormat(this.getDatePattern()).parse(value);
		} catch (e) {
			this.addError(errorKey, errorMessage);
		}
	};
	valid.validateDate = function(field, min, max, errorKey, errorMessage) {
		if (!this.isCheckedField(field)) return;
		delete valid.error[errorKey];
		var value = this.getParameter(field);
		if (value == null || "" == value.replace(/\s/ig, "")) {
			this.addError(errorKey, errorMessage);
			return ;
		}
		try {
			var format = new SimpleDateFormat(this.getDatePattern());
			var time = format.parse(value).getTime();
			if (time < format.parse(min).getTime() || time > format.parse(max).getTime())
				this.addError(errorKey, errorMessage);
		} catch (e) {
			this.addError(errorKey, errorMessage);
		}
	};
	valid.validateEqualField = function(field_1, field_2, errorKey, errorMessage) {
		if (!this.isCheckedField(field_2)) return;
		delete valid.error[errorKey];
		var value_1 = this.getParameter(field_1);
		var value_2 = this.getParameter(field_2);
		if (value_1 == null || value_2 == null || value_1 != value_2)
			this.addError(errorKey, errorMessage);
	};
	valid.validateEqualString = function(s1, s2, errorKey, errorMessage) {
		delete valid.error[errorKey];
		if (s1 == null || s2 == null || s1 != s2)
			this.addError(errorKey, errorMessage);
	};
	valid.validateEqualInteger = function(i1, i2, errorKey, errorMessage) {
		delete valid.error[errorKey];
		this.validateEqualString(i1, i2, errorKey, errorMessage);
	};
	valid.validateEmail = function(field, errorKey, errorMessage) {
		if (!this.isCheckedField(field)) return;
		delete valid.error[errorKey];
		this.validateRegex(field, emailAddressPattern, false, errorKey, errorMessage);
	};
	valid.validateEmailMany = function(field, errorKey, errorMessage) {
		if (!this.isCheckedField(field)) return;
		delete valid.error[errorKey];
		var value = this.getParameter(field);
		if (value == null || "" == value.replace(/\s/ig, "")) {
			this.addError(errorKey, errorMessage);
			return ;
		}
		var reg = new RegExp(emailAddressManyPattern, "i");
		if (!reg.test(/;$/.test(value) ? value : value + ";")) {
			this.addError(errorKey, errorMessage);
		}
	};
	valid.validateMobile = function(field, errorKey, errorMessage) {
		if (!this.isCheckedField(field)) return;
		delete valid.error[errorKey];
		this.validateRegex(field, mobileNumberPattern, false, errorKey, errorMessage);
	};
	valid.validateUrl = function(field, errorKey, errorMessage) {
		if (!this.isCheckedField(field)) return;
		delete valid.error[errorKey];
		this.validateRegex(field, urlAddressPattern, false, errorKey, errorMessage);
	};
	valid.validateRegex = function(field, regExpression, isCaseSensitive, errorKey, errorMessage) {
		if (!this.isCheckedField(field)) return;
		delete valid.error[errorKey];
		var value = this.getParameter(field);
		if (value == null || "" == value.replace(/\s/ig, "")) {
			this.addError(errorKey, errorMessage);
			return ;
		}
		if (typeof(isCaseSensitive) != "boolean" && errorMessage == undefined) {
			errorMessage = errorKey;
			errorKey = isCaseSensitive;
			isCaseSensitive = true;
		}
		var reg = isCaseSensitive ? new RegExp(regExpression) : new RegExp(regExpression, "i");
		if (!reg.test(value)) {
			this.addError(errorKey, errorMessage);
		}
	};
	valid.validateString = function(field, minLen, maxLen, errorKey, errorMessage) {
		if (!this.isCheckedField(field)) return;
		delete valid.error[errorKey];
		var value = this.getParameter(field);
		if (value == null || "" == value.replace(/\s/ig, "")) {
			this.addError(errorKey, errorMessage);
			return ;
		}
		if (value.length < minLen || value.length > maxLen)
			this.addError(errorKey, errorMessage);
	};
	valid.validateBoolean = function(field, errorKey, errorMessage) {
		if (!this.isCheckedField(field)) return;
		delete valid.error[errorKey];
		var value = this.getParameter(field);
		if (value == null || "" == value.replace(/\s/ig, "")) {
			this.addError(errorKey, errorMessage);
			return ;
		}
		value = value.replace(/\s/ig, "").toLowerCase();
		if ("1" == value || "true" == value)
			return ;
		else if ("0" == value || "false" == value)
			return ;
		this.addError(errorKey, errorMessage);
	};
	return valid;
}

(function($){
    $.fn.extend({
        validator: function(arg1, arg2, arg3, arg4){
            var selector = this.selector;
            // default parameter
        	var parameter = {forthwith: false};
        	for(var temp in Validate) {
    			parameter.name = temp;
    			break;
    		}
        	// checked parameter
        	var args = [arg1, arg2, arg3, arg4], flag;
        	for (var i in args) {
				switch (typeof(args[i])) {
				case "string":
					parameter.name = args[i];
					break;
				case "boolean":
					parameter.forthwith = args[i];
					break;
				case "function":
					if (flag) parameter.handlerError = args[i];
					else parameter.success = args[i];
					flag = true;
					break;
				}
			}
        	var validator = new Validate[parameter.name](selector, parameter.handlerError);
        	if (parameter.forthwith) {
        		var checkFunc = function () {
        			try {
        				validator.setCheckedField(this.name);
                        validator.validate();
                    } catch (e) { }
                    if(validator.invalid) {
                        try {
                            validator.handlerError(validator.error);
                        } catch (e) { }
                    }
        		};
        		var widget = $(selector).find("input[type=text], input[type=password], textarea");
        		widget.keyup(checkFunc); widget.blur(checkFunc);
			}
            $(this.selector).submit(function(){
            	try {
					validator.invalid = false;
            		validator.setCheckedField(null);
            		validator.validate();
            	} catch (e) { }
            	if(validator.invalid) {
            		try {
            			validator.handlerError(validator.error);
            		} catch (e) { }
            		return false;
            	}
            	var handler = false;
            	try {
            		handler = parameter.success(parameter.handlerError);
            	} catch (e) { }
            	if(handler != undefined && handler == false);
            		return false;
            });
        }
	});
})(jQuery);

File: validator-core/src/main/java/org/devefx/validator/core/Validator.java
package org.devefx.validator.core;

import java.io.IOException;
import java.io.InputStream;
import java.io.PrintWriter;
import java.io.Writer;
import java.net.MalformedURLException;
import java.net.URL;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

import org.devefx.validator.common.JsonResult;
import org.devefx.validator.common.StringUtils;

public abstract class Validator {
	
	private HttpServletRequest request;
	private HttpServletResponse response;
	private boolean invalid = false;
	private boolean shortCircuit = false;
	private String datePattern = null;
	private boolean scriptMode = false;
	private List<String> javascript = new ArrayList<String>();
	private String scriptText;
	private Map<String, String> error = new HashMap<String, String>();
	
	private static final String DEFAULT_DATE_PATTERN = "yyyy-MM-dd";
	private static final String emailAddressPattern = "\\b(^['_A-Za-z0-9-]+(\\.['_A-Za-z0-9-]+)*@([A-Za-z0-9-])+(\\.[A-Za-z0-9-]+)*((\\.[A-Za-z0-9]{2,})|(\\.[A-Za-z0-9]{2,}\\.[A-Za-z0-9]{2,}))$)\\b";
	private static final String mobileNumberPattern = "^(0|86|17951)?(13[0-9]|15[0-9]|17[678]|18[0-9]|14[57])[0-9]{8}$";
	
	protected void setShortCircuit(boolean shortCircuit) {
		if (scriptMode) {
			addScript(String.format("setShortCircuit(%b);", shortCircuit));
			return;
		}
		this.shortCircuit = shortCircuit;
	}
	
	protected void setDatePattern(String datePattern) {
		if (scriptMode) {
			addScript(String.format("setDatePattern('%s');", datePattern));
			return;
		}
		this.datePattern = datePattern;
	}
	
	protected String getDatePattern() {
		return (datePattern != null ? datePattern : DEFAULT_DATE_PATTERN);
	}
	
	protected Map<String, String> getError() {
		return error;
	}
	
	private void addScript(String script) {
		javascript.add(script);
	}
	
	protected void render(String viewName) {
		try {
			response.sendRedirect(viewName);
		} catch (IOException e) {
			e.printStackTrace();
		}
	}
	
	protected void renderJSONError() {
		JsonResult jsonResult = new JsonResult();
		jsonResult.put("success", !this.invalid);
		jsonResult.put("error", getError());
		try {
			response.setContentType("text/html;charset=UTF-8");
			PrintWriter out = response.getWriter();
			out.print(jsonResult.toString());
			out.flush();
		} catch (Exception e) {
			e.printStackTrace();
		}
	}
	
	public void output(Writer out) {
		try {
			if (this.scriptText == null) {
				Validator validator = getClass().newInstance();
				validator.scriptMode = true;
				validator.validate(null, null);

				String token = "";
				StringBuffer script = new StringBuffer();
				for (String s : validator.javascript) {
					script.append(token);
					script.append("\t\t\tthis.");
					script.append(s);
					token = "\n";
				}
				Map<String, Object> parameter = new HashMap<String, Object>();
				parameter.put("script", script.toString());
				parameter.put("name", getClass().getSimpleName());
				
				InputStream is = getClass().getClassLoader().getResourceAsStream("validator.vm");
				this.scriptText = StringUtils.format(StringUtils.reader(is, "utf-8"), parameter);
			}
		} catch (InstantiationException e) {
			e.printStackTrace();
		} catch (IllegalAccessException e) {
			e.printStackTrace();
		}
		
		PrintWriter printWriter = (PrintWriter) out;
		printWriter.write(scriptText);
		printWriter.flush();
	}
	
	final public boolean process(HttpServletRequest request, HttpServletResponse response) {
		this.request = request;
		this.response = response;
		try {
			this.validate(request, response);
		} catch (ValidateException e) {}
		if (this.invalid) {
			this.handleError(request, response);
			return false;
		}
		return true;
	}
	
	protected abstract void validate(HttpServletRequest request, HttpServletResponse response);
	
	protected abstract void handleError(HttpServletRequest request, HttpServletResponse response);
	
	private void addError(String errorKey, String errorMessage) {
		invalid = true;
		if (!error.containsKey(errorKey)) {
			request.setAttribute(errorKey, errorMessage);
			error.put(errorKey, errorMessage);
		}
		if (shortCircuit) {
			throw new ValidateException();
		}
	}
	
	protected void validateRequired(String field, String errorKey, String errorMessage) {
		if (scriptMode) {
			addScript(String.format("validateRequired('%s', '%s', '%s');", field, errorKey, errorMessage));
			return;
		}
		String value = request.getParameter(field);
		if (value == null || "".equals(value))
			addError(errorKey, errorMessage);
	}
	
	protected void validateRequiredString(String field, String errorKey, String errorMessage) {
		if (scriptMode) {
			addScript(String.format("validateRequiredString('%s', '%s', '%s');", field, errorKey, errorMessage));
			return;
		}
		String value = request.getParameter(field);
		if (value == null || "".equals(value.trim()))
			addError(errorKey, errorMessage);
	}
	
	protected void validateInteger(String field, int min, int max, String errorKey, String errorMessage) {
		if (scriptMode) {
			addScript(String.format("validateInteger('%s', %d, %d, '%s', '%s');", field, min, max, errorKey, errorMessage));
			return;
		}
		String value = request.getParameter(field);
		if (value == null || "".equals(value.trim())) {
			addError(errorKey, errorMessage);
			return ;
		}
		try {
			int temp = Integer.parseInt(value.trim());
			if (temp < min || temp > max)
				addError(errorKey, errorMessage);
		} catch (Exception e) {
			addError(errorKey, errorMessage);
		}
	}
	
	protected void validateInteger(String field, String errorKey, String errorMessage) {
		if (scriptMode) {
			addScript(String.format("validateInteger('%s', '%s', '%s');", field, errorKey, errorMessage));
			return;
		}
		String value = request.getParameter(field);
		if (value == null || "".equals(value.trim())) {
			addError(errorKey, errorMessage);
			return ;
		}
		try {
			Integer.parseInt(value.trim());
		} catch (Exception e) {
			addError(errorKey, errorMessage);
		}
	}
	
	protected void validateLong(String field, long min, long max, String errorKey, String errorMessage) {
		if (scriptMode) {
			addScript(String.format("validateLong('%s', %d, %d, '%s', '%s');", field, min, max, errorKey, errorMessage));
			return;
		}
		String value = request.getParameter(field);
		if (value == null || "".equals(value.trim())) {
			addError(errorKey, errorMessage);
			return ;
		}
		try {
			long temp = Long.parseLong(value.trim());
			if (temp < min || temp > max)
				addError(errorKey, errorMessage);
		} catch (Exception e) {
			addError(errorKey, errorMessage);
		}
	}
	
	protected void validateLong(String field, String errorKey, String errorMessage) {
		if (scriptMode) {
			addScript(String.format("validateLong('%s', '%s', '%s');", field, errorKey, errorMessage));
			return;
		}
		String value = request.getParameter(field);
		if (value == null || "".equals(value.trim())) {
			addError(errorKey, errorMessage);
			return ;
		}
		try {
			Long.parseLong(value.trim());
		} catch (Exception e) {
			addError(errorKey, errorMessage);
		}
	}
	
	protected void validateDouble(String field, double min, double max, String errorKey, String errorMessage) {
		if (scriptMode) {
			addScript(String.format("validateDouble('%s', %d, %d, '%s', '%s');", field, min, max, errorKey, errorMessage));
			return;
		}
		String value = request.getParameter(field);
		if (value == null || "".equals(value.trim())) {
			addError(errorKey, errorMessage);
			return ;
		}
		try {
			double temp = Double.parseDouble(value.trim());
			if (temp < min || temp > max)
				addError(errorKey, errorMessage);
		} catch (Exception e) {
			addError(errorKey, errorMessage);
		}
	}
	
	protected void validateDouble(String field, String errorKey, String errorMessage) {
		if (scriptMode) {
			addScript(String.format("validateDouble('%s', '%s', '%s');", field, errorKey, errorMessage));
			return;
		}
		String value = request.getParameter(field);
		if (value == null || "".equals(value.trim())) {
			addError(errorKey, errorMessage);
			return ;
		}
		try {
			Double.parseDouble(value.trim());
		} catch (Exception e) {
			addError(errorKey, errorMessage);
		}
	}
	
	protected void validateDate(String field, String errorKey, String errorMessage) {
		if (scriptMode) {
			addScript(String.format("validateDate('%s', '%s', '%s');", field, errorKey, errorMessage));
			return;
		}
		String value = request.getParameter(field);
		if (value == null || "".equals(value.trim())) {
			addError(errorKey, errorMessage);
			return ;
		}
		try {
			new SimpleDateFormat(getDatePattern()).parse(value.trim());
		} catch (Exception e) {
			addError(errorKey, errorMessage);
		}
	}
	
	protected void validateDate(String field, Date min, Date max, String errorKey, String errorMessage) {
		if (scriptMode) {
			SimpleDateFormat sdf = new SimpleDateFormat(getDatePattern());
			addScript(String.format("validateDate('%s', '%s', '%s', '%s', '%s');", field, sdf.format(min), sdf.format(max), errorKey, errorMessage));
			return;
		}
		String value = request.getParameter(field);
		if (value == null || "".equals(value.trim())) {
			addError(errorKey, errorMessage);
			return ;
		}
		try {
			Date temp = new SimpleDateFormat(getDatePattern()).parse(value.trim());
			if (temp.before(min) || temp.after(max))
				addError(errorKey, errorMessage);
		} catch (Exception e) {
			addError(errorKey, errorMessage);
		}
	}
	
	protected void validateDate(String field, String min, String max, String errorKey, String errorMessage) {
		if (scriptMode) {
			addScript(String.format("validateDate('%s', '%s', '%s', '%s', '%s');", field, min, max, errorKey, errorMessage));
			return;
		}
		try {
			SimpleDateFormat sdf = new SimpleDateFormat(getDatePattern());
			validateDate(field, sdf.parse(min.trim()), sdf.parse(max.trim()), errorKey, errorMessage);
		} catch (Exception e) {
			addError(errorKey, errorMessage);
		}
	}
	
	protected void validateEqualField(String field_1, String field_2, String errorKey, String errorMessage) {
		if (scriptMode) {
			addScript(String.format("validateEqualField('%s', '%s', '%s', '%s');", field_1, field_2, errorKey, errorMessage));
			return;
		}
		String value_1 = request.getParameter(field_1);
		String value_2 = request.getParameter(field_2);
		if (value_1 == null || value_2 == null || (! value_1.equals(value_2)))
			addError(errorKey, errorMessage);
	}
	
	protected void validateEqualString(String s1, String s2, String errorKey, String errorMessage) {
		if (scriptMode) {
			addScript(String.format("validateEqualString('%s', '%s', '%s', '%s');", s1, s2, errorKey, errorMessage));
			return;
		}
		if (s1 == null || s2 == null || (! s1.equals(s2)))
			addError(errorKey, errorMessage);
	}
	
	protected void validateEqualInteger(Integer i1, Integer i2, String errorKey, String errorMessage) {
		if (scriptMode) {
			addScript(String.format("validateEqualInteger(%d, %d, '%s', '%s');", i1, i2, errorKey, errorMessage));
			return;
		}
		if (i1 == null || i2 == null || (i1.intValue() != i2.intValue()))
			addError(errorKey, errorMessage);
	}
	
	protected void validateEmail(String field, String errorKey, String errorMessage) {
		if (scriptMode) {
			addScript(String.format("validateEmail('%s', '%s', '%s');", field, errorKey, errorMessage));
			return;
		}
		validateRegex(field, emailAddressPattern, false, errorKey, errorMessage);
	}
	
	protected void validateMobile(String field, String errorKey, String errorMessage) {
		if (scriptMode) {
			addScript(String.format("validateMobile('%s', '%s', '%s');", field, errorKey, errorMessage));
			return;
		}
		validateRegex(field, mobileNumberPattern, false, errorKey, errorMessage);
	}
	
	protected void validateUrl(String field, String errorKey, String errorMessage) {
		if (scriptMode) {
			addScript(String.format("validateUrl('%s', '%s', '%s');", field, errorKey, errorMessage));
			return;
		}
		String value = request.getParameter(field);
        if (value == null) {
        	addError(errorKey, errorMessage);
        	return ;
        }
        try {
			value = value.trim();
			if (value.startsWith("https://"))
				value = "http://" + value.substring(8);
			new URL(value);
		} catch (MalformedURLException e) {
			addError(errorKey, errorMessage);
		}
	}
	
	protected void validateRegex(String field, String regExpression, boolean isCaseSensitive, String errorKey, String errorMessage) {
		if (scriptMode) {
			addScript(String.format("validateRegex('%s', '%s', %b, '%s', '%s');", field, regExpression, isCaseSensitive, errorKey, errorMessage));
			return;
		}
		String value = request.getParameter(field);
        if (value == null) {
        	addError(errorKey, errorMessage);
        	return ;
        }
        Pattern pattern = isCaseSensitive ? Pattern.compile(regExpression) : Pattern.compile(regExpression, Pattern.CASE_INSENSITIVE);
        Matcher matcher = pattern.matcher(value);
        if (!matcher.matches())
        	addError(errorKey, errorMessage);
	}
	
	protected void validateRegex(String field, String regExpression, String errorKey, String errorMessage) {
		validateRegex(field, regExpression, true, errorKey, errorMessage);
	}
	
	protected void validateString(String field, int minLen, int maxLen, String errorKey, String errorMessage) {
		if (scriptMode) {
			addScript(String.format("validateString('%s', %d, %d, '%s', '%s');", field, minLen, maxLen, errorKey, errorMessage));
			return;
		}
		String value = request.getParameter(field);
        if (value == null) {
        	addError(errorKey, errorMessage);
        	return ;
        }
        if (value.length() < minLen || value.length() > maxLen)
			addError(errorKey, errorMessage);
	}
	
	protected void validateBoolean(String field, String errorKey, String errorMessage) {
		if (scriptMode) {
			addScript(String.format("validateBoolean('%s', '%s', '%s');", field, errorKey, errorMessage));
			return;
		}
		String value = request.getParameter(field);
        if (value == null) {
        	addError(errorKey, errorMessage);
        	return ;
        }
        value = value.trim().toLowerCase();
		if ("1".equals(value) || "true".equals(value))
			return ;
		else if ("0".equals(value) || "false".equals(value))
			return ;
		addError(errorKey, errorMessage);
	}
	
	/*protected void validateToken(String tokenName, String errorKey, String errorMessage) {
		
	}
	
	protected void validateToken(String errorKey, String errorMessage) {
		
	}*/
}
